llvm/include/llvm/Support/PropertySetIO.h

```diff
@@ -0,0 +1,163 @@
+  template <typename T> static Type getTypeTag();
+
+  // Casts from int value to a type tag.
+  static Expected<Type> getTypeTag(int T) {
+    if (T < first || T > last)
+      return createStringError(std::error_code(), "bad property type " + T);
+    return static_cast<Type>(T);
+  }
+
+  PropertyValue() : Ty(NONE) {}
+  PropertyValue(Type T) : Ty(T) {}
+
+  PropertyValue(uint32_t Val) : Ty(UINT32), Val({Val}) {}
+  PropertyValue(const PropertyValue &P) : Ty(P.Ty), Val({P.Val}) {}
+  PropertyValue(PropertyValue &&P) : Ty(P.Ty), Val({P.Val}) {}
+
+  PropertyValue &operator=(PropertyValue &&P) = default;
+
+  PropertyValue &operator=(const PropertyValue &P) = default;
+
+  // get property value as unsigned 32-bit integer
+  uint32_t asUint32() const {
+    assert(Ty == UINT32);
+    return Val.UInt32Val;
+  }
+
+  bool isValid() const { return getType() != NONE; }
+
+  // set property value; the 'T' type must be convertible to a property type tag
+  template <typename T> void set(T V) {
+    assert(getTypeTag<T>() == Ty);
+    getValueRef<T>() = V;
+  }
+
+  Type getType() const { return Ty; }
+
+  size_t size() const {
+    switch (Ty) {
+    case UINT32:
+      return sizeof(Val.UInt32Val);
+    default:
+      llvm_unreachable_internal("unsupported property type");
+    }
+  }
+
+private:
+  template <typename T> T &getValueRef();
+
+  Type Ty;
+  union {
+    uint32_t UInt32Val;
+  } Val;
+};
+
+std::ostream &operator<<(std::ostream &Out, const PropertyValue &V);
+
+// A property set. Preserves insertion order when iterating elements.
+using PropertySet = MapVector<StringRef, PropertyValue>;
+
+// A "registry" of multiple property sets. Maps a property set name to its
+// contents. Can be read/written.
+class PropertySetRegistry {
+public:
+  // using MapTy = std::map<StringRef, PropertySet>;
+  using MapTy = MapVector<StringRef, PropertySet>;
+
+  // Specific property category names used by tools.
+  static constexpr char SYCL_SPECIALIZATION_CONSTANTS[] =
+      "SYCL/specialization constants";
+
+  // Function for bulk addition of an entire property set under given category
+  // (property set name).
+  template <typename T>
+  void add(StringRef Category, const std::map<StringRef, T> &Props) {
+    assert(PropSetMap.find(Category) == PropSetMap.end() &&
+           "category already added");
+    auto &PropSet = PropSetMap[Category];
+
+    for (const auto &Prop : Props)
+      PropSet.insert(std::make_pair(Prop.first, PropertyValue(Prop.second)));
+  }
+
+  // Parses and creates a property set registry.
+  static Expected<std::unique_ptr<PropertySetRegistry>> read(MemoryBuffer *Buf);
+
+  // Dumps a property set registry to a stream.
+  void write(raw_ostream &Out) const;
+
+  // Start iterator of all preperty sets in the registry.
+  const auto begin() { return PropSetMap.begin(); }
+  // End iterator of all preperty sets in the registry.
+  const auto end() { return PropSetMap.end(); }
```

Comment:
I suggest to use explicit type here instead of auto (i.e. MapTy::const_iterator). And these methods need const qualified.

Reply:
ok